base/math.jl

```diff
@@ -187,7 +187,22 @@ function evalpoly(x, p::Tuple)
     end
 end
 
-evalpoly(x, p::AbstractVector) = _evalpoly(x, p)
+function evalpoly(x, p::AbstractVector)
+    i = lastindex(p)
+
+    # without the annotation v is not type stable in some cases,
+    # e.g. if p contains Float32 but x is Float64
+    T = promote_type(typeof(x), eltype(p))
+    @inbounds v::T = p[i]
```

Comment:
Instead, you could do `v  = p[i] * one(eltype(p))`.  The compiler will optimize out the multiplication for hardware-floating-point types, and it will give you the type you want.

Reply:
This `@inbounds` is not correct, because you haven't checked that `p` is nonempty.

Probably we should have a line:
```jl
isempty(p) && return zero(x) * one(eltype(p))
```
(The current `evalpoly` throws in this case.)